pkg/elemental/elemental.go

```diff
@@ -58,64 +62,63 @@ func (e *Elemental) FormatPartition(part *v1.Partition, opts ...string) error {
 // required partitions
 func (e *Elemental) PartitionAndFormatDevice(i v1.SharedInstallSpec) error {
-	disk := partitioner.NewDisk(
-		i.GetTarget(),
-		partitioner.WithRunner(e.config.Runner),
-		partitioner.WithFS(e.config.Fs),
-		partitioner.WithLogger(e.config.Logger),
-	)
-
-	if !disk.Exists() {
+	if _, err := os.Stat(i.GetTarget()); os.IsNotExist(err) {
 		e.config.Logger.Errorf("Disk %s does not exist", i.GetTarget())
 		return fmt.Errorf("disk %s does not exist", i.GetTarget())
 	}
 
-	e.config.Logger.Infof("Partitioning device...")
-	out, err := disk.NewPartitionTable(i.GetPartTable())
+	disk, err := partitioner.NewDisk(i.GetTarget(), partitioner.WithLogger(e.config.Logger))
 	if err != nil {
-		e.config.Logger.Errorf("Failed creating new partition table: %s", out)
 		return err
 	}
 
-	parts := i.GetPartitions().PartitionsByInstallOrder(i.GetExtraPartitions())
-	return e.createPartitions(disk, parts)
-}
-
-func (e *Elemental) createAndFormatPartition(disk *partitioner.Disk, part *v1.Partition) error {
-	e.config.Logger.Debugf("Adding partition %s", part.Name)
-	num, err := disk.AddPartition(part.Size, part.FS, part.Name, part.Flags...)
-	if err != nil {
-		e.config.Logger.Errorf("Failed creating %s partition", part.Name)
-		return err
-	}
-	partDev, err := disk.FindPartitionDevice(num)
+	e.config.Logger.Infof("Partitioning device...")
+	err = disk.NewPartitionTable(i.GetPartTable(), i.GetPartitions().PartitionsByInstallOrder(i.GetExtraPartitions()))
 	if err != nil {
+		e.config.Logger.Errorf("Failed creating new partition table: %s", err)
 		return err
 	}
-	if part.FS != "" {
-		e.config.Logger.Debugf("Formatting partition with label %s", part.FilesystemLabel)
-		err = partitioner.FormatDevice(e.config.Runner, partDev, part.FS, part.FilesystemLabel)
-		if err != nil {
-			e.config.Logger.Errorf("Failed formatting partition %s", part.Name)
-			return err
-		}
-	} else {
-		e.config.Logger.Debugf("Wipe file system on %s", part.Name)
-		err = disk.WipeFsOnPartition(partDev)
+
+	// Only re-read table on devices. On files there is no need and this call will fail
+	if disk.Type == diskfs.Device {
+		err = disk.ReReadPartitionTable()
 		if err != nil {
-			e.config.Logger.Errorf("Failed to wipe filesystem of partition %s", partDev)
+			e.config.Logger.Errorf("Reread table: %s", err)
 			return err
 		}
 	}
-	part.Path = partDev
-	return nil
-}
 
-func (e *Elemental) createPartitions(disk *partitioner.Disk, parts v1.PartitionList) error {
-	for _, part := range parts {
-		err := e.createAndFormatPartition(disk, part)
-		if err != nil {
-			return err
+	table, err := disk.GetPartitionTable()
+	if err != nil {
+		e.config.Logger.Errorf("table: %s", err)
+		return err
+	}
+	err = disk.Close()
+	if err != nil {
+		e.config.Logger.Errorf("Close disk: %s", err)
+	}
+	// Sync changes
+	syscall.Sync()
+	// Trigger udevadm to refresh devices
+	_, _ = e.config.Runner.Run("udevadm", "trigger")
+	_, _ = e.config.Runner.Run("udevadm", "settle")
```

Comment:
is it worth putting the output of these calls? or at least something in the log to say doing trigger now, doing settle now, etc?

Reply:
I dont think so, the calls actually return nothing, no output from the commands so there is no value on it I think. I never ever seen those error out as well xD

Reply:
Since this function already returns an error, I don't see why we shouldn't handle the error here. It may never happen but if it does, it makes it a real pita to debug with zero-output and error being skipped. To my experience, skipping errors never proved to be a good idea.